### PR TITLE
Form Upgrades 1.x

### PR DESCRIPTION
**Pull request by claytondaley**
This PR implements improvements from the @grizzm0 branch exemplar (see #576) to the 1.x branch of ZfcUser.  This PR leaves in events in 1.x for backwards compatibility.  For a comparable 2.x branch, I plan to fully eliminate them to be consistent with the discussion in #576.

The shift to `FormElementManager` raises a variety of points worth noting (and possibly discussing):
- I feel that it's important to make it clear that the underlying Forms now depend on `FormElementManager` to be fully initialized.  My gut feeling is that ZF really needs to automatically handle Forms through the FEM (even if they're mistakenly listed in the `service_manager`).  In the meantime, I moved the factories to a dedicated namespace `FormElementManagerFactory` to make the dependency more explicit.  I don't love this approach and am open to suggestions.
- This PR could introduce backwards compatibility issues for user who have inherited from ZfcUser forms, but are not themselves using `FormElementManager` to get them (in these cases `init()` will not be called so elements won't be created).
  - ZfcUserAdmin will definitely be impacted.  I'll try to extract a matching PR from my work on that project on the expectation that they'll be committed at the same time.

### CODE REVIEW

**Comment by Danielss89** [comment]
We can't really introduce any BC breaks in 1.X branch. Unless it's security related.

**Comment by claytondaley** [comment]
Where exactly is a line drawn on BC?  I could wrap this in some FEM-aware logic that ensures (both at the Factory and Constructor level) that `init()` is run if the caller is not the FEM.

If that's too risky, I don't mind if we simply ditch it.  This was more of a PoC for me since my app was using ZfcUserAdmin so it was easier (for me) to test improvements on the 1.x branch.

**Comment by Danielss89** [comment]
BC break is if a user upgrades ZfcUser and his application errors because of changes in ZfcUser.

**Comment by claytondaley** [comment]
I understand conceptually what BC means, but a user's implementation could depend on almost any arbitrary implementation detail.  You'd paralyze yourself if you took this idea to an extreme.

For example, the current 1.x doesn't `add()` the submit button [in the ChangeEmailForm](https://github.com/ZF-Commons/ZfcUser/pull/578/files#diff-4b982965105bba771198a291a1676803R59).  Instead, it's hard-coded in the [change email view](https://github.com/ZF-Commons/ZfcUser/pull/578/files#diff-02c9c26a18a5c1830e7d46bf561eb22dL34).  This is inconsistent with the rest of ZfcUser in  way that's disruptive to extension/usage.

"Fixing" 1.x by moving the button to the constructor would be transparent to someone using ZfcUser out-of-the-box.  However, anyone who replaced that view might have replicated the hard-coded submit button.  For that user it wouldn't be backwards-compatible.

One answer is "it needs to pass all the tests without modification".  That's fair and (likely) easily done.  A more absolute answer would suggest that no changes should be allowed period except in the case of security.

**Comment by Danielss89** [comment]
If the submit is hardcoded it current version, and you add it to the form, it won't break for other users when updating because they didn't use it anyway, so it's an addition, not really a change, and that's acceptable.

If it was the other way around, and the user showed the button by formSubmit(), and you removed it from the form object, then it would error, and that would be a BC break, which we can't have in 1.x.

**Comment by claytondaley** [comment]
Right.  Unless the new view builds the form by iterating all of the elements and calling `formElement()`.  Or adds a submit button (using a different name) and uses either a looped `formElement()` or `form()` view helper in their view. In both cases, a duplicate submit button appears.

Please understand that I agree with your sentiment and am not trying to be difficult.  Development isn't my profession and I've only made tangential contributions to open source projects.  This is the first time I've had to negotiate the BC line and I could probably articulate a "failing" case for pretty much any non-trivial change.

**Comment by claytondaley** [comment]
OK.  I rolled back all of my modified tests to evaluate backwards compatibility.  I had to fix a couple mocks to support added calls to `FormElementManager`, but otherwise the code now passes the original tests:

The good:

 - All the closures in `Module.php` have been pulled out into factories.
 - Forms should now support custom objects pretty robustly:
     - All Forms are loaded internally using `FormElementManager`.  If you pointed `zfcuser_login_form` at a custom object that has elements in `init()` (the FEM pattern),  they should load properly.
     - Separately, the factories proactively inject `FormElementManager` into forms.  If you just `get('zfcuser_login_form')` and then `->add()` a custom element (for example, in a Delegator), it should also work.
 - All ZfcUser forms now use the generic form rendering engine
     - I fixed some quirks like the place where submit buttons are added

The bad:

 - I gave up on moving element initialization for ZfcUser forms into `init()` (the FEM pattern).
      - This became too complex because backwards compatibility requires that constructors call `init()` if they're not being loaded by FEM.
      - I had a working technique (rolled back in [b1b6835](https://github.com/claytondaley/ZfcUser/commit/b1b6835d05d6dce97a951732660a3c6db6f8203c)) that was reasonably friendly to inheritance, but it was ugly, risked odd side-effects, and was unnecessary since none of the built-in elements are actually custom.
      - The one downside is that an inheriting class can't just override `init()` to replace base elements (presumably with custom elements or `construct()` would be enough).  It's a minor hassle and nothing new.

**Comment by stijnhau** [comment]
Anyn progress on this?

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Created by .ignore support plugin (hsz.mobi)
```
Comment (Danielss89) [inline .gitignore]: No need to polute the .ignore file with all this. Please revert to previous version.

File: .travis.yml [hunk 2]
```diff
@@ -21,10 +21,6 @@ script:
 after_script:
   - php ocular.phar code-coverage:upload --format=php-clover build/logs/clover.xml
 
-notifications:
```
Comment (Danielss89) [inline .travis.yml]: Readd this

File: view/zfc-user/user/changeemail.phtml [hunk 3]
```diff
@@ -11,25 +11,6 @@ $form = $this->changeEmailForm;
...
-        <?php endif ?>
-    <?php endforeach ?>
-    </dl>
-    <?php if ($this->redirect): ?>
-        <input type="hidden" name="redirect" value="<?php echo $this->redirect ?>" />
-    <?php endif ?>
-    <input type="submit" value="Submit" />
-<?php echo $this->form()->closeTag() ?>
+
+<?php include '_form.phtml'; ?>
```
Comment (Danielss89) [inline view/zfc-user/user/changeemail.phtml]: Please use a partial instead

File: view/zfc-user/user/changepassword.phtml [hunk 4]
```diff
@@ -14,34 +14,6 @@ $form->setAttribute('method', 'post');
...
-
-<dd><?php
-//echo $this->formInput($form->get('csrf'));
-echo $this->formInput($form->get('submit'));
-?></dd>
-
-</dl>
 
-<?php echo $this->form()->closeTag(); ?>
+<?php include '_form.phtml'; ?>
```
Comment (Danielss89) [inline view/zfc-user/user/changepassword.phtml]: Use partial

File: view/zfc-user/user/login.phtml [hunk 5]
```diff
@@ -7,26 +7,8 @@ $form->setAttribute('action', $this->url('zfcuser/login'));
...
-
-        <?php if ($this->redirect): ?>
-            <input type="hidden" name="redirect" value="<?php echo $this->escapeHtml($this->redirect) ?>" />
-        <?php endif ?>
-
-        <dd><?php echo $this->formButton($form->get('submit')) ?></dd>
-    </dl>
-
-<?php echo $this->form()->closeTag() ?>
+<?php include '_form.phtml'; ?>
```
Comment (Danielss89) [inline view/zfc-user/user/login.phtml]: Use partial

File: view/zfc-user/user/register.phtml [hunk 6]
```diff
@@ -11,22 +11,4 @@ $form->setAttribute('action', $this->url('zfcuser/register'));
...
-        <?php else: ?>
-            <dd><?php echo $this->formInput($element) . $this->formElementErrors($element) ?></dd>
-        <?php endif ?>
-    <?php endforeach ?>
-    </dl>
-    <?php if ($this->redirect): ?>
-        <input type="hidden" name="redirect" value="<?php echo $this->escapeHtml($this->redirect) ?>" />
-    <?php endif ?>
-<?php echo $this->form()->closeTag() ?>
+<?php include '_form.phtml'; ?>
```
Comment (Danielss89) [inline view/zfc-user/user/register.phtml]: Use partial

File: src/ZfcUser/Factory/Form/ChangeEmail.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+use Zend\ServiceManager\FactoryInterface;
+use Zend\ServiceManager\ServiceLocatorInterface;
+use ZfcUser\Form;
+use ZfcUser\Validator;
+
+class ChangeEmail implements FactoryInterface
+{
+    public function createService(ServiceLocatorInterface $formElementManager)
+    {
+        if ($formElementManager instanceof FormElementManager) {
```
Comment (Danielss89) [inline src/ZfcUser/Factory/Form/ChangeEmail.php]: If the element is registered in formElementManager it's always the formElementManager you'll get injected, so i'm not sure i see the use for this `if` check?

Comment (claytondaley) [inline src/ZfcUser/Factory/Form/ChangeEmail.php]: Legacy code doesn't necessarily construct the class from the FEM.  We need to check if the ServiceLocator is actually the FEM or just the regular ServiceManager.  In the latter case, we need to take the initiative to obtain the FEM.

**Comment by claytondaley** [comment]
I'll revert the other miscellaneous issues (gitignore and travis).  What's the logic behind the partials?

**Comment by Danielss89** [comment]
Partial is ZF's way to include files, with the correct renderer, and you can pass in variables too.

**Comment by stijnhau** [comment]
Maybe a new milestone should be added bacause it isn't in the just realeased version 1.3.3

**Comment by claytondaley** [comment]
@Danielss89, I think I was confused by your original request and want to make sure I'm on the same page now.

This line (and logic) was copied from [ZfcUserAdmin](https://github.com/Danielss89/ZfcUserAdmin/blob/master/view/zfc-user-admin/user-admin/create.phtml#L10) because I figured your project was doing it the desired way.  I think you're asking me to replace all of these instances with a line like:
```
<?php echo $this->partial('_form.phtml'); ?>
```
If so, that's an easy change and I can do it.  If you're asking for something else, I need clarification.

**Comment by Danielss89** [comment]
@stijnhau ahhhr fuck, was supposed to be 1.2.3 :S

**Comment by Danielss89** [comment]
@claytondaley Yes exactly. ZfcUserAdmin is old, so don't look at that for best practices :P

**Comment by stijnhau** [comment]
@Danielss89 
Maybe even delete the version?

**Comment by Danielss89** [comment]
@stijnhau nop. 1.3.3 is the new version.
The old 1.3.0 is now 1.4.0

**Comment by claytondaley** [comment]
Sorry for the delay.  Partials should all be added.

**Pull request merged**